plugin/cache/cache.go

```diff
@@ -0,0 +1,87 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+// New creates a cacher.
+func New() *Cache {
+	return &Cache{
+		data: map[string]*cacheEntry{},
+	}
+}
+
+// Func is the signature for a cache function.
+type Func func() (interface{}, error)
+
+// Cache is the internal cache implementation.
+type Cache struct {
+	lock sync.RWMutex
+	data map[string]*cacheEntry
+}
+
+// cacheEntry represents an item in the cache with an expiration and lifetime.
+type cacheEntry struct {
+	result   interface{}
+	created  time.Time
+	lifetime time.Duration
+}
+
+// Fetch retrieves an item from the cache. If the item exists in the cache and
+// is within its lifetime, it is returned. If the item does not exist, or if the
+// item exists but has exceeded its lifetime, the function f is invoked and the
+// result is updated in the cache.
+func (c *Cache) Fetch(name string, t time.Duration, f Func) (interface{}, error) {
+	// Attempt to read from the cache, returning the cached value if it's still
+	// valid.
+	c.lock.RLock()
+	e, ok := c.data[name]
+	if ok && e.result != nil && time.Now().UTC().Sub(e.created) < e.lifetime {
```

Comment:
The `UTC()` conversion isn't needed in these calculations (or the time storage).